make sure gen_backend_stubs is tested in CI

bdhirsh opened the pull request:
From the commend here, I realized `test_gen_backend_stubs.py` hasn't been getting run in CI.

There was a change a few weeks ago (someone added a global cache) that caused the test to start failing locally, so I fixed it and added it to the list of CI tests in this PR.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61296 make sure gen_backend_stubs is tested in CI**
* #61295 add basic regex test to detect invalid CompositeExplicitAutograd ops that are implemented with DispatchStub
* #61133 fix eig.e bug - it uses DispatchStub and therefore isn't backend agnostic

Differential Revision: [D29561064](https://our.internmc.facebook.com/intern/diff/D29561064)

bdhirsh committed:
make sure gen_backend_stubs is tested in CI

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29561064).

ezyang commented on test/test_gen_backend_stubs.py:
```diff
@@ -28,13 +29,18 @@ def get_errors_from_gen_backend_stubs(self, yaml_str: str) -> str:
 
+    def cleanup(self) -> None:
+        # the codegen has a global cache that we need to reset between tests
+        tools.codegen.gen._GLOBAL_PARSE_NATIVE_YAML_CACHE.clear()
```
This seems bad. cc @albanD 

It would have been better if @soulitzer's suggestion had been applied; then the cache would only be applied when `native_functions.yaml` is parsed (refactor to call a single function) and direct parsing as is done here would bypass.

ezyang replied:
If you just want to stem the bleeding in this patch, clearing the cache in `def setUp` is preferable to hard coding it in all the sites.

albanD replied:
Ho sorry about that... I wasn't expecting the codegen to used outside of a standalone script.
We can definitely make this design better.

ezyang replied:
@albanD now you know. Also when it doesn't cost that much to handle the generalizations (as was the case here; soulitzer's alternative is essentially the same length) it is better to put them in.

bdhirsh replied:
I'm not sure that @functools.cache would fix this (I think that was soulitzer's suggestion?). Each call to `gen_backend_stubs.py:run()` from a unit test re-invokes `parse_native_yaml()` [here](https://github.com/pytorch/pytorch/blob/d7b31fe95da98a2798376044dc84e5f9b7a3e486/tools/codegen/gen_backend_stubs.py#L189), and they all return the same cached NamedTuple object.

`gen_backend_stubs.py:run()` ends up modifying that object - specifically, it adds new keys to the `backend_indices` dictionary in the named tuple (https://github.com/pytorch/pytorch/blob/d7b31fe95da98a2798376044dc84e5f9b7a3e486/tools/codegen/gen_backend_stubs.py#L83) - which will modify the cached object that will be returned on the next invocation.

Maybe I should make the return type of `parse_native_yaml()` a frozen dataclass instead of a NamedTuple, and enforce that we have to copy the dictionary instead of mutating it. I can confirm that it doesn't slow down the build by too much (hopefully it won't - we're copying a few thousand BackendMetadata objects, each of which are small).

Do you think it's worth making the return type a frozen dataclass, and having gen_backend_stubs.py

ezyang replied:
Yes, yes, if we use functools cache we have to make the function purely functional. Not a big deal, IMO.

bdhirsh committed:
Update on "make sure gen_backend_stubs is tested in CI"

From the commend here, I realized `test_gen_backend_stubs.py` hasn't been getting run in CI.

There was a change a few weeks ago (someone added a global cache) that caused the test to start failing locally, so I fixed it and added it to the list of CI tests in this PR.


Differential Revision: [D29561064](https://our.internmc.facebook.com/intern/diff/D29561064)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29561064).

bdhirsh committed:
Update on "make sure gen_backend_stubs is tested in CI"

From the commend here, I realized `test_gen_backend_stubs.py` hasn't been getting run in CI.

There was a change a few weeks ago (someone added a global cache) that caused the test to start failing locally, so I fixed it and added it to the list of CI tests in this PR.


Differential Revision: [D29561064](https://our.internmc.facebook.com/intern/diff/D29561064)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29561064).
